numpy/lib/shape_base.py

```diff
@@ -16,9 +16,112 @@
 __all__ = [
     'column_stack', 'row_stack', 'dstack', 'array_split', 'split',
     'hsplit', 'vsplit', 'dsplit', 'apply_over_axes', 'expand_dims',
-    'apply_along_axis', 'kron', 'tile', 'get_array_wrap'
+    'apply_along_axis', 'kron', 'tile', 'get_array_wrap', 'take_along_axis'
     ]
 
+def take_along_axis(arr, indices, axis):
+    """
+    Take the elements described by `indices` along each 1-D slice of the given
+    `axis`, matching up subspaces of arr and indices.
+
+    This function can be used to index with the result of `argsort`, `argmax`,
+    and other functions.
+
+    This computes::
+
+        out[a..., k..., b...] = arr[a..., indices[a..., k..., b...], b...]
+
+    .. versionadded:: 1.13.0
+
+    Parameters
+    ----------
+    arr: array_like (A..., M, B...)
+        source array
+    indices: array_like (A..., K..., B...)
+        indices to take along each 1d slice of `arr`
+    axis: int
+        the axis to take 1d slices along
+
+    Returns
+    -------
+    out: ndarray (A..., K..., B...)
+        The indexed result.
+
+    See Also
+    --------
+    take : Take along an axis without matching up subspaces
+
+    Examples
+    --------
+
+    For this sample array
+
+    >>> a = np.array([[10, 30, 20], [60, 40, 50]])
+
+    We can sort either by using sort directly, or argsort and this function
+
+    >>> np.sort(a, axis=1)
+    array([[10, 20, 30],
+           [40, 50, 60]])
+    >>> ai = np.argsort(a, axis=1); ai
+    array([[0, 2, 1],
+           [1, 2, 0]], dtype=int64)
+    >>> np.take_along_axis(a, ai, axis=1)
+    array([[10, 20, 30],
+           [40, 50, 60]])
+
+    The same works for max and min:
+
+    >>> np.max(a, axis=1)
+    array([30, 60])
+    >>> ai = np.argmax(a, axis=1); ai
+    array([1, 0], dtype=int64)
+    >>> np.take_along_axis(a, ai, axis=1)
```

Comment:
I think the problem here comes down to where extra `(1,)` dimensions should be inserted, and I'm inclined to agree with @seberg that there are too many options, and forcing the user to specify is best.

So the rule would become

```python
assert(ind.ndim == x.ndim)
out[i..., :, j...] = x[i..., :, j...][ind[i..., :, j...]
```

Which would mean that you would write (`taa = take_along_axis`):

```python
# if you really want to squeeze that axis
max = taa(a, np.expand_dims(a.argmax(axis=axis), axis=axis), axis=axis).squeeze(axis=axis)
assert_equal(max, a.max(axis=axis))

# but lets be honest, squeezing is a bad plan for broadcasting
maxk = taa(a, a.argmax(axis=axis, keep_dims=True), axis=axis)
assert_equal(max, a.max(axis=axis, keep_dims=True))

# sort just works
sort = taa(a, a.argsort(axis=axis), axis=axis)
assert_equal(sort, a.sort(axis=axis, True))

# if you want to add another dimension, you make yourself explicit about where you want it
a_ext = np.expand_dims(a, axis=-1)
ai1 = np.argsort(a_ext, axis=axis )
ai2 = np.argsort(abs(a_ext), axis=axis )
ai = np.concatenate((ai1, ai2), axis=-1)
sort_a_abs_a = taa(a_ext, ai, axis=axis)

assert_equal(sort_a_abs_a[...,0], a_ext.sort(axis=axis))
assert_equal(sort_a_abs_a[...,1], abs(a_ext).sort(axis=axis))
```